### PR TITLE
feat: add StaticParsedLine

### PR DESCRIPTION
**Pull request by tustvold**
In order to aggregate lines across requests it is necessary for them to own the data they refer to, so they can be stored in a collection, etc...

This rejigs the line splitting logic to return index ranges, instead of string slices, to then allow ouroboros to work its magic and construct StaticParsedLine that retain a reference to the originating `Arc<str>`. This formalation preserves the zero-copy semantics of the parser.

Unfortunately there is [an issue](https://github.com/servo/rust-smallvec/issues/217) with smallvec that causes it to not be correctly covariant. This PR therefore removes SmallVec, especially since moving the jemalloc I imagine the difference will be largely negligible and the parser is certainly not the ingest bottleneck at the moment. Smallvec v2, whenever it is released, should fix this bug.

### CODE REVIEW

File: influxdb_line_protocol/src/lib.rs [hunk 1]
```diff
@@ -1136,25 +1185,18 @@ mod test {
...
+        assert_eq!(trim_line("  foo  boo", 1, 8).unwrap(), (2, 8));
+        assert_eq!(trim_line("  foo  boo", 3, 8).unwrap(), (3, 8));
+        assert!(trim_line("  foo  boo", 0, 2).is_none());
+        assert!(trim_line("  foo  boo", 5, 7).is_none());
     }
 
     #[test]
     fn test_split_lines() {
-        assert_eq!(split_lines("").collect::<Vec<_>>(), vec![""]);
+        assert_eq!(split_lines("").count(), 0);
```
Comment (tustvold) [inline influxdb_line_protocol/src/lib.rs:1199]: Split lines now also trims the returned strings

File: influxdb_line_protocol/src/lib.rs [hunk 2]
```diff
@@ -525,49 +565,62 @@ fn split_lines(input: &str) -> impl Iterator<Item = &str> {
...
-            return false;
-        }
+    let mut last_idx = 0;
+    let mut chars = input.chars().enumerate();
 
-        if c == ' ' {
-            fields = true;
-            return false;
-        }
+    std::iter::from_fn(move || {
```
Comment (tustvold) [inline influxdb_line_protocol/src/lib.rs:572]: Unfortunately the pattern API used internally by `split` is experimental, so have to DIY it a bit

File: influxdb_line_protocol/src/lib.rs [hunk 3]
```diff
@@ -525,49 +565,62 @@ fn split_lines(input: &str) -> impl Iterator<Item = &str> {
...
+        for (idx, c) in &mut chars {
```
[the remark this answers is not in the document]
Comment (tustvold) [reply influxdb_line_protocol/src/lib.rs:573]: Nice - will fix :+1:

File: influxdb_line_protocol/src/lib.rs [hunk 4]
```diff
@@ -758,22 +816,28 @@ fn field_bool_value(i: &str) -> IResult<&str, bool> {
...
-            i = &i[offset..];
-        } else {
-            break i;
+// Trims a single line slice returning None if this is a comment
+// or has no non-whitespace characters
+fn trim_line(i: &str, start: usize, end: usize) -> Option<(usize, usize)> {
+    let str = &i[start..end];
+    let i = str.find(|c: char| !c.is_whitespace())?;
+
+    // Unfortunately rfind only returns the start byte index
```
Comment (tustvold) [inline influxdb_line_protocol/src/lib.rs:825]: If someone knows how to do this in a nicer way I'd be interested to know

**Comment by alamb** [comment]
@tustvold  is there a writeup somewhere that describes what you plan to do with this code (you mention something about "aggregating across requests")?

I find I can offer much better reviews of code if I understand the usecase (aka the *why* are you making this change) which I don't feel like I do quite yet.

File: influxdb_line_protocol/src/lib.rs [hunk 5]
```diff
@@ -228,6 +231,30 @@ impl<'a> Display for ParsedLine<'a> {
...
 }
 
+/// A `ParsedLine` that owns its data
+#[self_referencing]
+pub struct StaticParsedLine {
+    data: Arc<str>,
+
+    #[borrows(data)]
+    #[covariant]
+    line: ParsedLine<'this>,
```
Comment (alamb) [inline influxdb_line_protocol/src/lib.rs:241]:
```suggestion
    /// `ParsedLine` backed by a structure that points to some part of `data`
    line: ParsedLine<'this>,
```

FWIW I wonder if having a Vec<ParsedLine<'this>> is what you will eventually want (so you can pass batches of parsed lines around without having to allocate for the `Arc` each time.

File: influxdb_line_protocol/src/lib.rs [hunk 6]
```diff
@@ -228,6 +231,30 @@ impl<'a> Display for ParsedLine<'a> {
     }
 }
 
+/// A `ParsedLine` that owns its data
+#[self_referencing]
+pub struct StaticParsedLine {
```
Comment (alamb) [inline influxdb_line_protocol/src/lib.rs:236]: I found the name `Static` confusing (as this isn't `'static` data). Perhaps it would be better named  `ParsedLineAndData` or `OwnedParsedLine` 🤔 .

File: influxdb_line_protocol/src/lib.rs [hunk 7]
```diff
@@ -525,7 +565,7 @@ fn split_lines(input: &str) -> impl Iterator<Item = &str> {
     let mut commas = 0;
 
     let mut in_escape = false;
-    input.split(move |c| {
+    let is_delimiter = move |c| {
```
Comment (alamb) [inline influxdb_line_protocol/src/lib.rs:568]: This function is on the hot path (when I was messing around with it previously). I recommend running the `line_parser` benchmark

**Review by alamb — CHANGES_REQUESTED** [review]
Now that I have studied this more I see what this is doing and I (really) like it. The main benefit of `StaticParsedLine` effectively (in my mind) allows parsing to happen in one set of tasks and consumption of those lines in another. I think this would lead to better structured code 

However, on my machine, the performance benchmarks via `cargo bench -- line_parser ( first on this branch and then on main) suggest this branch is ~ 50% slower than main (performance improved on main compared to this branch).  It is probably worth double checking my numbers, but I don't think we should merge this if it causes such a performance regression. 



```
line_parser/all lines   time:   [1.2173 ms 1.2228 ms 1.2290 ms]
                        thrpt:  [90.337 MiB/s 90.790 MiB/s 91.203 MiB/s]
                 change:
                        time:   [-37.756% -37.090% -36.452%] (p = 0.00 < 0.05)
                        thrpt:  [+57.362% +58.956% +60.657%]
                        Performance has improved.
```

I don't know how much of that was due to the loss of smallvec vs the other changes to the hot code path. 

I wonder if you could avoid the changes to the hot path by having `StaticParsedLine` contain a `Vec` of `ParsedLines` and collect the result of `parsed_lines`

So in summary: 
* Idea 👍 
* Performance regression 👎

**Comment by tustvold** [comment]
@alamb I'm struggling to reproduce your findings...

On main

```
running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 82 filtered out; finished in 0.00s

line_parser/all lines   time:   [880.93 us 881.55 us 882.31 us]                                  
                        thrpt:  [125.83 MiB/s 125.94 MiB/s 126.03 MiB/s]
```

On this branch

```
line_parser/all lines   time:   [869.83 us 870.20 us 870.63 us]                                  
                        thrpt:  [127.52 MiB/s 127.58 MiB/s 127.64 MiB/s]
                 change:
                        time:   [-1.5726% -1.4175% -1.3014%] (p = 0.00 < 0.05)
                        thrpt:  [+1.3185% +1.4378% +1.5977%]
                        Performance has improved.
```

File: influxdb_line_protocol/src/lib.rs (hunk 5 above)
Comment (tustvold) [reply influxdb_line_protocol/src/lib.rs:241]: I specifically need to be able to split per-line, into different IOx shards, and then subsequently within the producer

File: influxdb_line_protocol/src/lib.rs (hunk 6 above)
Comment (tustvold) [reply influxdb_line_protocol/src/lib.rs]: I guess I was thinking that it's effectively a ParsedLine<'static> but yeah, OwnedParsedLine is more obvious

File: influxdb_line_protocol/src/lib.rs (hunk 7 above)
Comment (tustvold) [reply influxdb_line_protocol/src/lib.rs:571]: This code is effectively unchanged, the only major difference is that Rust's implementation uses unsafe slice access, but when playing around the difference seems to be pretty marginal at least on my system.

**Comment by alamb** [comment]
I reran the benchmark (on a macbook pro, btw) and I still see a slowdown. 

Specifically, I ran:

On main @ ebea554e5a9fd591efdcb32880b67ce639150092 `cargo bench -- line_parser`

Then on this branch @ 3d210bb8459e2df4e3ca1d44261940eef212229d  `cargo bench -p influxdb_line_protocol -- line_parser`

```
line_parser/all lines   time:   [1.9917 ms 2.0127 ms 2.0337 ms]                                  
                        thrpt:  [54.591 MiB/s 55.162 MiB/s 55.742 MiB/s]
                 change:
                        time:   [+55.166% +57.180% +59.220%] (p = 0.00 < 0.05)
                        thrpt:  [-37.194% -36.378% -35.553%]
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe
```

🤔 

(amusingly my benchmark reports speeds almost 2x slower in absolute magnitude than your machine :) )

File: influxdb_line_protocol/Cargo.toml [hunk 8]
```diff
@@ -6,9 +6,14 @@ edition = "2018"
...
 snafu = "0.6.2"
 observability_deps = { path = "../observability_deps" }
+ouroboros = "0.8.3"
 
 [dev-dependencies] # In alphabetical order
+criterion = "0.3.3"
 test_helpers = { path = "../test_helpers" }
+
+[[bench]]
+name = "line_parser"
```
Comment (alamb) [inline influxdb_line_protocol/Cargo.toml:18]: The bench was moved! This threw me for a loop while trying to re-run the benchmarks

However, this is a nice change

Comment (tustvold) [reply influxdb_line_protocol/Cargo.toml:18]: Yeah, I didn't like building the server crate every time I wanted to run it :laughing:

**Comment by tustvold** [comment]
@alamb I've pushed a branch https://github.com/influxdata/influxdb_iox/tree/owned-lines-jemalloc that uses jemalloc instead of the system allocator, could you give that a try?

**Comment by alamb** [comment]
> that uses jemalloc instead of the system allocator, could you give that a try? 
... trying ...

**Comment by e-dard** [comment]
Here is what I see:

**macOS intel i9**
```
⇒  critcmp main owned-lines
group                                                                          main                                     owned-lines
-----                                                                          ----                                     -----------
line_parser/all lines                                                          1.00  1219.6±13.87µs    91.0 MB/sec      1.54  1877.9±37.19µs    59.1 MB/sec
```

**macOS M1 (`aarch64`)
```
```

So for both of these there is a slow down. I think the linux behaviour is probably more important but we should get to the bottom of it because the regression on these architectures is so significant.

**Comment by alamb** [comment]
Here is the comparison of main to `owned-lines-jemalloc`:. It still shows a slowdown but much less pronounced (10% rather than 50%)

```
line_parser/all lines   time:   [1.3603 ms 1.3650 ms 1.3704 ms]                                  
                        thrpt:  [81.015 MiB/s 81.333 MiB/s 81.614 MiB/s]
                 change:
                        time:   [+11.125% +11.826% +12.716%] (p = 0.00 < 0.05)
                        thrpt:  [-11.281% -10.576% -10.011%]
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe
```

**Comment by tustvold** [comment]
I've pushed a commit that adds a replacement for Smallvec, or at least the functionality of it we need. Testing shows this to perform on-par with smallvec, but without the covariance issue. It does, however, use a small amount of unsafe - this makes about a ~8% difference to the benchmark, I will be interested to hear what people think about this.

On my Linux machine this branch now represents a pretty consistent speedup of ~2%. On mac the story is somewhat more complex. I'm testing on an Intel Mac Mini and the benchmark performance is rather inconsistent, it is common to see swings of 12% on the same code and I've struggled to find a consistent testing methodology - it almost seems like the system doesn't ramp up to fullspeed until it has seen a consistent load for a period, but then if it has seen a consistent load for too long it starts to throttle. The benchmark results are below, I'm inclined to call it probably equivalent.

Tests were run with `cargo bench` on this branch and `line-protocol-bench` (which moves the benchmark to avoid having to compile the server crate)

```
Mac

line-protocol-bench

line_parser/all lines   time:   [1.4093 ms 1.4163 ms 1.4244 ms]                                  
                        thrpt:  [77.945 MiB/s 78.388 MiB/s 78.777 MiB/s]


line_parser/all lines   time:   [1.4064 ms 1.4098 ms 1.4139 ms]                                  
                        thrpt:  [78.522 MiB/s 78.748 MiB/s 78.942 MiB/s]
                        
line_parser/all lines   time:   [1.2621 ms 1.2653 ms 1.2693 ms]                                  
                        thrpt:  [87.466 MiB/s 87.743 MiB/s 87.965 MiB/s]

line_parser/all lines   time:   [1.4104 ms 1.4171 ms 1.4245 ms]                                  
                        thrpt:  [77.936 MiB/s 78.344 MiB/s 78.718 MiB/s]

line_parser/all lines   time:   [1.4093 ms 1.4155 ms 1.4227 ms]                                  
                        thrpt:  [78.039 MiB/s 78.432 MiB/s 78.776 MiB/s]


owned-lines


line_parser/all lines   time:   [1.3666 ms 1.3706 ms 1.3750 ms]                                  
                        thrpt:  [80.742 MiB/s 81.004 MiB/s 81.237 MiB/s]

line_parser/all lines   time:   [1.3698 ms 1.3749 ms 1.3807 ms]                                  
                        thrpt:  [80.409 MiB/s 80.748 MiB/s 81.053 MiB/s]


line_parser/all lines   time:   [1.3727 ms 1.3787 ms 1.3852 ms]                                  
                        thrpt:  [80.151 MiB/s 80.529 MiB/s 80.878 MiB/s]


line_parser/all lines   time:   [1.3693 ms 1.3759 ms 1.3833 ms]                                  
                        thrpt:  [80.260 MiB/s 80.691 MiB/s 81.078 MiB/s]


Linux

line-protocol-bench

line_parser/all lines   time:   [872.83 us 873.34 us 874.13 us]                                  
                        thrpt:  [127.01 MiB/s 127.12 MiB/s 127.20 MiB/s]

line_parser/all lines   time:   [898.19 us 898.72 us 899.17 us]                                  
                        thrpt:  [123.47 MiB/s 123.53 MiB/s 123.61 MiB/s]

line_parser/all lines   time:   [897.22 us 897.84 us 898.39 us]                                  
                        thrpt:  [123.58 MiB/s 123.65 MiB/s 123.74 MiB/s]

owned-lines

line_parser/all lines   time:   [851.61 us 852.13 us 852.99 us]                                  
                        thrpt:  [130.16 MiB/s 130.29 MiB/s 130.37 MiB/s]

line_parser/all lines   time:   [870.45 us 870.58 us 870.69 us]                                  
                        thrpt:  [127.51 MiB/s 127.53 MiB/s 127.55 MiB/s]

line_parser/all lines   time:   [860.31 us 860.51 us 860.74 us]                                  
                        thrpt:  [128.98 MiB/s 129.02 MiB/s 129.05 MiB/s]
```

**Comment by tustvold** [comment]
Following on from today's meeting, kinesis/kafka producer support is not on the near term roadmap - so not going to look to get these changes in at this time
